blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +267,225 @@ where
+                        return Some(Err(Error::Other(error_msg)))
+                    }
+                    let error_msg =format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: `Error Converting message to Cid` n{:})",cid,message.sequence(),s.sequence());
+                    return Some(Err(Error::Other(error_msg)))
+                }
+                if s.sequence() < message.sequence() {
+                    return Some(Ok(None));
+                }
+
+                None
+            })
+            .next()
+            .unwrap_or_else(|| Ok(None))
+    }
+
+    /// WaitForMessage blocks until a message appears on chain. It looks backwards in the chain to see if this has already
+    /// happened. It guarantees that the message has been on chain for at least confidence epochs without being reverted
+    /// before returning.
+    pub async fn wait_for_message(
+        &self,
+        cid: &Cid,
+        confidence: i64,
+    ) -> Result<Option<(Arc<Tipset>, MessageReceipt)>, Error> {
+        let mut subscribers = self.subscriber.clone().ok_or_else(|| {
+            Error::Other("State Manager not subscribed to tipset head changes".to_string())
+        })?;
+        let mut back_search_wait = self.back_search_wait.clone().ok_or_else(|| {
+            Error::Other("State manager not subscribed to back search wait".to_string())
+        })?;
+        let message = chain::get_chain_message(self.get_block_store_ref(), cid)
+            .map_err(|err| Error::Other(format!("failed to load message {:}", err)))?;
+
+        let maybe_subscriber: Option<HeadChange> = subscribers.next().await;
+        let first_subscriber = maybe_subscriber.ok_or_else(|| {
+            Error::Other("SubHeadChanges first entry should have been one item".to_string())
+        })?;
+
+        let tipset = match first_subscriber {
+            HeadChange::Current(tipset) => tipset,
+            _ => {
+                return Err(Error::Other(format!(
+                    "expected current head on SHC stream (got {:?})",
+                    first_subscriber
+                )))
+            }
+        };
+        let maybe_message_reciept = self.tipset_executed_message(&tipset, cid, &*message)?;
+        if let Some(r) = maybe_message_reciept {
+            return Ok(Some((tipset.clone(), r)));
+        }
+
+        let (back_tipset, _back_receipt) = self
+            .search_back_for_message(&tipset, &*message)?
+            .ok_or_else(|| {
+                Error::Other("State manager not subscribed to back search wait".to_string())
+            })?;
+
+        let mut candidate_tipset: Option<Arc<Tipset>> = None;
+        let mut candidate_receipt: Option<MessageReceipt> = None;
+        let height_of_head = tipset.epoch();
+        let mut reverts: HashMap<TipsetKeys, bool> = HashMap::new();
+        loop {
```

Comment:
Isn't this an infinite loop? how does this get broken out of?

Reply:
They are return closes under Apply ,and back_search_wait variants

Reply:
Looks like there is an outer for in the lotus implementation as well 

https://github.com/filecoin-project/lotus/blob/c4de2373fd13c909876cf459e59ead7749b9939d/chain/stmgr/stmgr.go#L538

Reply:
nah that's just one loop, you have a loop within a loop, if the while let ever comes up with None, it will be an infinite loop? Seems sketchy to rely on breaks for this

Reply:
Oh yeahh was confused cause I saw a for in a case from the go side and read it wrong to think it would always check the case until it finds something. I should remove it